### PR TITLE
Assert observedGeneration is incremented in Status.Conditions.

### PR DESCRIPTION
**Pull request by dprotaso**
**What type of PR is this?**

/kind cleanup
/kind api-change

**What this PR does / why we need it**:
This PR ensures Gateway Resources and their conditions update observedGeneration as described in the GEP.

Current it tests
- [x] Gateway
- [x] GatewayClass
- [x] HTTPRoute

**Which issue(s) this PR fixes**:

Fixes: https://github.com/kubernetes-sigs/gateway-api/issues/1584

**Does this PR introduce a user-facing change?**:

```release-note
Conformance tests now assert observedGeneration is incremented in Status.Conditions (GatewayClass, Gateway, HTTPRoute)
```

### CODE REVIEW

**Comment by dprotaso** [comment]
Also question: How come there are no existing conformance tests for GatewayClass mutations?

**Comment by robscott** [comment]
>Also question: How come there are no existing conformance tests for GatewayClass mutations?

What would that look like? The conformance tests accept a GatewayClass to run tests against, but I can't think of anything meaningful to actually test on a GatewayClass other than that it is accepted.

**Comment by dprotaso** [comment]
> but I can't think of anything meaningful to actually test on a GatewayClass other than that it is accepted.

Until now with ensuring `observedGeneration` gets bumped.

Thus changing the `parameterRef` or even `description` should trigger this

File: conformance/utils/kubernetes/helpers.go
Comment (youngnick) [inline conformance/utils/kubernetes/helpers.go:40]: Feels like a paste error maybe? 😂

**Review by youngnick — COMMENTED** [review]
The approach LGTM. This could go into v0.6.0, or a v0.6.1 update later if need be.

File: conformance/utils/kubernetes/helpers.go
Comment (dprotaso) [reply conformance/utils/kubernetes/helpers.go:40]: lol that's hilarious

File: conformance/tests/gateway-observed-generation-bump.go
Comment (dprotaso) [inline conformance/tests/gateway-observed-generation-bump.go:65]: nit: this listener port already exists - should use `8080`

File: conformance/tests/gateway-observed-generation-bump.go
Comment (dprotaso) [inline conformance/tests/gateway-observed-generation-bump.go:45]: nit: timeout of 20 seconds might be too short

File: conformance/tests/gatewayclass-observed-generation-bump.go
Comment (dprotaso) [inline conformance/tests/gatewayclass-observed-generation-bump.go:45]: increase timeout

File: conformance/tests/httproute-observed-generation-bump.go
Comment (dprotaso) [inline conformance/tests/httproute-observed-generation-bump.go:54]: Increase timeout

File: conformance/tests/httproute-observed-generation-bump.yaml
Comment (dprotaso) [inline conformance/tests/httproute-observed-generation-bump.yaml:11]: non-existent backed

File: conformance/tests/httproute-observed-generation-bump.go
Comment (dprotaso) [inline conformance/tests/httproute-observed-generation-bump.go:69]: non-existent backed

**Comment by dprotaso** [comment]
This is ready for review - I haven't tested this against any implementations yet. I'll do my own testing but I don't think that should get the review.

If any implementors are willing to help test this change that would be much appreciated.

File: conformance/tests/gateway-observed-generation-bump.yaml
Comment (robscott) [inline conformance/tests/gateway-observed-generation-bump.yaml:4]: Any way to use an existing Gateway for this?

File: conformance/tests/gatewayclass-observed-generation-bump.go
Comment (robscott) [inline conformance/tests/gatewayclass-observed-generation-bump.go:36]: The GatewayClass is part of the environment provided by the person running the conformance tests. I personally don't think we should modify it. Maybe we can ensure that ObservedGeneration in status of the provided GatewayClass matches the generation of the resource though?

File: conformance/tests/httproute-observed-generation-bump.go
Comment (robscott) [inline conformance/tests/httproute-observed-generation-bump.go:91]: This feels like it should cause the test to fail

File: conformance/tests/httproute-observed-generation-bump.go
Comment (robscott) [inline conformance/tests/httproute-observed-generation-bump.go:38]: Can we make this sufficiently generic that the included code can work for every route type? This can be a follow up, just would be nice to have consistency across all Route types.

File: conformance/tests/httproute-observed-generation-bump.go
Comment (robscott) [inline conformance/tests/httproute-observed-generation-bump.go:66]: This applies to all the comments like this - the wording/pluralization feels a bit off here. Maybe `Expected observedGeneration to be updated for all conditions, only x/y were updated.`

**Review by robscott — COMMENTED** [review]
Thanks @dprotaso!

File: conformance/tests/gateway-observed-generation-bump.go
Comment (sunjayBhatia) [inline conformance/tests/gateway-observed-generation-bump.go:42]: ran tests w/ Contour, so far ran into the fact that this name is incorrect, should be `gateway-observed-generation-bump` (or some existing Gateway if possible, similar to other comments)

File: conformance/tests/gateway-observed-generation-bump.go
Comment (sunjayBhatia) [inline conformance/tests/gateway-observed-generation-bump.go:56]: shouldn't this be as below? (here and in all other similar places)
```suggestion
			if !kubernetes.GatewayConditionsHaveLatestObservedGeneration(existing) {
```

File: conformance/tests/httproute-observed-generation-bump.go
Comment (sunjayBhatia) [inline conformance/tests/httproute-observed-generation-bump.go:84]: I dont think we can do it this way, `Update` will modify the generation of the local variable

if you print the generation before/after the update calls, e.g.

```
			fmt.Println("original route", existing.GetGeneration())

			// Sanity check
			if !kubernetes.HTTPRouteConditionsHaveLatestObservedGeneration(existing) {
				t.Fatal("Not all the condition's observedGeneration were updated")
			}

			existing.Spec.Rules[0].BackendRefs[0].Name = "infra-backend-v2"
			err = s.Client.Update(ctx, existing)
			require.NoErrorf(t, err, "error updating the HTTPRoute: %v", err)

			fmt.Println("original route after update", existing.GetGeneration())

			kubernetes.HTTPRouteMustHaveCondition(t, s.Client, s.TimeoutConfig, routeNN, gwNN, acceptedCondition)

			updated := &v1beta1.HTTPRoute{}
			err = s.Client.Get(ctx, routeNN, updated)
			require.NoErrorf(t, err, "error getting Gateway: %v", err)

			fmt.Println("updated route", updated.GetGeneration())
```

I get 

```
original route 1
original route after update 2
obj generation 2
```

so we should capture the generation of the original resource before we update it and not rely on fetching it from `existing` later

File: conformance/tests/httproute-observed-generation-bump.go
Comment (sunjayBhatia) [inline conformance/tests/httproute-observed-generation-bump.go:89]: This check never works when I run it because I always get an empty list of parent statuses coming from `existing`, I think we need to examine exactly what state of the route ends up in the `expected` object since `Update` will update the content that `expected` points to

```
	// Update updates the given obj in the Kubernetes cluster. obj must be a
	// struct pointer so that obj can be updated with the content returned by the Server.
	Update(ctx context.Context, obj Object, opts ...UpdateOption) error
```

can we change this to just assert on the exact set of parent refs we expect? since we know what the route is we also know what the parentrefs should be from the fixture

File: conformance/tests/httproute-observed-generation-bump.go
Comment (sunjayBhatia) [inline conformance/tests/httproute-observed-generation-bump.go:94]: would it be easier to just modify `HTTPRouteMustHaveCondition` to check generations?

File: conformance/tests/gateway-observed-generation-bump.go
Comment (sunjayBhatia) [inline conformance/tests/gateway-observed-generation-bump.go:65]: at the moment, contour will not mark this update as accepted, we don't support a port other than 80 or 443

Comment (sunjayBhatia) [reply conformance/tests/gateway-observed-generation-bump.go:65]: can we do something trivial here like you've done with gw class, or add an annotation etc.?

Comment (sunjayBhatia) [reply conformance/tests/gateway-observed-generation-bump.go:65]: or maybe change an existing listeners allowedroutes section?

File: conformance/tests/gatewayclass-observed-generation-bump.go
Comment (sunjayBhatia) [inline conformance/tests/gatewayclass-observed-generation-bump.go:38]: currently Contour is not reconciling changes to GatewayClasses, just taking their state when they are first created/observed by the Contour Gateway API provisioner due to:

https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.GatewayClass

```
It is recommended that this resource be used as a template for Gateways. This means that a Gateway is based on the state of the GatewayClass at the time it was created and changes to the GatewayClass or associated parameters are not propagated down to existing Gateways. This recommendation is intended to limit the blast radius of changes to GatewayClass or associated parameters. If implementations choose to propagate GatewayClass changes to existing Gateways, that MUST be clearly documented by the implementation.
```

We haven't yet finished our thinking/implementation on this area so this fails for us

File: conformance/utils/kubernetes/helpers.go
Comment (sunjayBhatia) [inline conformance/utils/kubernetes/helpers.go:108]: this and below can be simplified to `return ConditionsHaveLatestObservedGeneration(gc, gc.Status.Conditions)`

File: conformance/tests/httproute-observed-generation-bump.go
Comment (sunjayBhatia) [reply conformance/tests/httproute-observed-generation-bump.go:84]: similar for status, probably capture the status before we update the resource to use later

File: conformance/tests/gateway-observed-generation-bump.go
Comment (dprotaso) [reply conformance/tests/gateway-observed-generation-bump.go:65]: > we don't support a port other than 80 or 443

This restriction is odd - the API doesn't specify any constraints on the port numbers

File: conformance/tests/gateway-observed-generation-bump.yaml
Comment (dprotaso) [reply conformance/tests/gateway-observed-generation-bump.yaml:4]: I would prefer to keep tests isolated - I don't want to change the base gateway if other tests assume it's static.

File: conformance/tests/gatewayclass-observed-generation-bump.go
Comment (dprotaso) [reply conformance/tests/gatewayclass-observed-generation-bump.go:36]: > Maybe we can ensure that ObservedGeneration in status of the provided GatewayClass matches the generation of the resource though?

We do this already - this test is to ensure the observedGeneration changes when the `spec` changes

File: conformance/tests/gateway-observed-generation-bump.go
Comment (sunjayBhatia) [reply conformance/tests/gateway-observed-generation-bump.go:65]: well I may be misspeaking a little, but at the very least Contour doesn't let you configure multiple HTTP or HTTPS Listeners, only one port is allowed for each at the moment

File: conformance/tests/httproute-observed-generation-bump.go
Comment (dprotaso) [reply conformance/tests/httproute-observed-generation-bump.go:38]: Will do as a follow up

File: conformance/tests/httproute-observed-generation-bump.go
Comment (dprotaso) [reply conformance/tests/httproute-observed-generation-bump.go:94]: I already did but `HTTPRouteMustHaveCondition` polls until the conditions are in an acceptable state - I don't want to poll here.

File: conformance/tests/gatewayclass-observed-generation-bump.go
Comment (dprotaso) [reply conformance/tests/gatewayclass-observed-generation-bump.go:38]: Given that `GatewayClass` has an `Accepted` condition I would expect the observedGeneration of that condition to increment when the GWC spec changes.

File: conformance/tests/gateway-observed-generation-bump.go
Comment (dprotaso) [reply conformance/tests/gateway-observed-generation-bump.go:63]: Created an issue to move this discussion off the PR - https://github.com/kubernetes-sigs/gateway-api/issues/1607

File: conformance/tests/httproute-observed-generation-bump.go
Comment (dprotaso) [reply conformance/tests/httproute-observed-generation-bump.go]: Resolved - thanks for pointing this out

**Comment by dprotaso** [comment]
Hey folks - this is ready for another review.

File: conformance/utils/kubernetes/helpers.go
Comment (sunjayBhatia) [inline conformance/utils/kubernetes/helpers.go:502]: when looking at the test logs, these lines end up being hard to read, e.g.

```
=== RUN   TestConformance/HTTPRouteObservedGenerationBump/observedGeneration_should_increment
    helpers.go:209: Gateways and Pods in gateway-conformance-infra namespaces ready
    helpers.go:502: HTTPRoute(controller=projectcontour.io/gateway-controller,ref=v1beta1.RouteParentStatus{ParentRef:v1beta1.ParentReference{Group:(*v1beta1.Group)(0xc00041e310), Kind:(*v1beta1.Kind)(0xc00041e350), Namespace:(*v1beta1.Na
mespace)(nil), Name:"same-namespace", SectionName:(*v1beta1.SectionName)(nil), Port:(*v1beta1.PortNumber)(nil)}, ControllerName:"projectcontour.io/gateway-controller", Conditions:[]v1.Condition{v1.Condition{Type:"Accepted", Status:"True",
 ObservedGeneration:1, LastTransitionTime:time.Date(2022, time.December, 16, 17, 46, 3, 0, time.Local), Reason:"Accepted", Message:"Accepted HTTPRoute"}}}) expected observedGeneration to be updated for all conditions, only 0/1 were update
d. stale conditions are: Accepted
```

if we want to keep these logs might be good to pare this down a bit (and in other places the same pattern is used)

File: conformance/tests/gateway-observed-generation-bump.go
Comment (sunjayBhatia) [inline conformance/tests/gateway-observed-generation-bump.go:85]: could use `require.NotEqual` here

File: conformance/tests/gatewayclass-observed-generation-bump.yaml
Comment (robscott) [inline conformance/tests/gatewayclass-observed-generation-bump.yaml:6]: This seems like a significant change. As I understand it, this requires implementations to watch GatewayClasses matching their controller name and then update status on those. I think in some cases GatewayClass is part of the deployment model and each controller supports exactly one class. I could be wrong, but need to get some more input from other implementations before moving forward with this.

File: conformance/utils/suite/suite.go
Comment (robscott) [inline conformance/utils/suite/suite.go:149]: Seems like we need a docs change to go along with this if we're going to move forward with this approach.

**Review by robscott — COMMENTED** [review]
Thanks @dprotaso! I think the changes for Gateway and HTTPRoute are very reasonable/non-controversial and could likely merge in soon. I think we need to get more feedback from implementations to see if the proposed GWC test is actually implementable though. The alternative would be to modify the provided GatewayClass, but that does not seem ideal either. It _may_ be worth splitting out the GWC side of this since it could take longer to get in, but completely up to you.

**Comment by robscott** [comment]
Adding a hold until we can confirm the proposed GatewayClass tests are broadly implementable.

/hold

File: conformance/tests/gatewayclass-observed-generation-bump.yaml
Comment (dprotaso) [reply conformance/tests/gatewayclass-observed-generation-bump.yaml:6]: I can `t.Skip()` this test for now as not to hold the entire PR

File: conformance/utils/suite/suite.go
Comment (dprotaso) [reply conformance/utils/suite/suite.go:152]: done

File: conformance/tests/gatewayclass-observed-generation-bump.yaml
Comment (dprotaso) [reply conformance/tests/gatewayclass-observed-generation-bump.yaml:6]: I put it behind a support feature flag for now

File: conformance/utils/kubernetes/helpers.go
Comment (dprotaso) [reply conformance/utils/kubernetes/helpers.go]: Done - just using `namespace/name` of the parent ref

**Comment by dprotaso** [comment]
I created a support featuere `GatewayClassObservedGenerationBump` to allow implementations to opt into that test while it's being discussed.

File: conformance/tests/gateway-observed-generation-bump.go
Comment (sunjayBhatia) [reply conformance/tests/gateway-observed-generation-bump.go:65]: this test is passing now on Contour since observed generation is set correctly, the status we get is:

```
status:
  addresses:
  - type: IPAddress
    value: 172.24.255.203
  conditions:
  - lastTransitionTime: "2022-12-16T19:15:41Z"
    message: Gateway is accepted
    observedGeneration: 2
    reason: Accepted
    status: "True"
    type: Accepted
  - lastTransitionTime: "2022-12-16T19:15:41Z"
    message: Listeners are not valid
    observedGeneration: 2
    reason: ListenersNotValid
    status: "False"
    type: Programmed
  listeners:
  - attachedRoutes: 0
    conditions:
    - lastTransitionTime: "2022-12-16T19:15:41Z"
      message: Only one HTTP port is supported
      observedGeneration: 2
      reason: PortUnavailable
      status: "False"
      type: Accepted
    - lastTransitionTime: "2022-12-16T19:15:41Z"
      message: Invalid listener, see other listener conditions for details
      observedGeneration: 2
      reason: Invalid
      status: "False"
      type: Programmed
    name: alternate
    supportedKinds: []
  - attachedRoutes: 0
    conditions:
    - lastTransitionTime: "2022-12-16T19:15:41Z"
      message: Valid listener
      observedGeneration: 2
      reason: Programmed
      status: "True"
      type: Programmed
    name: http
    supportedKinds:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
```

The Gateway as a whole should still be `Accepted=true` since some part of the config is valid, per: https://gateway-api.sigs.k8s.io/geps/gep-1364/#accepted

observedGenerations are all correct, though the status/validity isn't all there

File: conformance/tests/gatewayclass-observed-generation-bump.go
Comment (sunjayBhatia) [reply conformance/tests/gatewayclass-observed-generation-bump.go:39]: ah nvm, found where we were short circuiting the status update logic, this should work with Contour we have a small fix to make

Comment (sunjayBhatia) [reply conformance/tests/gatewayclass-observed-generation-bump.go:39]: https://github.com/projectcontour/contour/pull/4932

**Comment by youngnick** [comment]
This seems good to me, especially with the opt-out for the GatewayClass for now, so I'm going to hit the approve button. I'll leave it to one of the other approvers to give it the tick.

/approve

**Review by shaneutt — APPROVED** [review]
In terms of implementability, @pmalek @mlavacca and myself all talked about this and don't have any concerns for the multiple GWAPI implementations @ Kong :+1:

File: conformance/utils/suite/suite.go
Comment (robscott) [reply conformance/utils/suite/suite.go:152]: To clarify, I was trying to say that we need an update to our user-facing docs here: https://github.com/kubernetes-sigs/gateway-api/blob/main/site-src/concepts/conformance.md#running-tests

Comment (robscott) [reply conformance/utils/suite/suite.go:152]: Actually misunderstood this, we're already reading this from GatewayClass resource so users don't need to configure this.

**Comment by robscott** [comment]
Hey @dprotaso, thanks for all your work on this! The tests LGTM, can you squash this PR into a single commit? (See https://kubernetes.io/docs/contribute/new-content/open-a-pr/#squashing-commits for more info). Once that's done, ping any of us for an LGTM and we'll get it merged.

**Comment by dprotaso** [comment]
Weird isn’t your prow configured to squash on merge?

**Comment by shaneutt** [comment]
> Weird isn’t your prow configured to squash on merge?

Yes, but I think the suggestion here is since this PR is quite large we would like to give you the opportunity to squash into _less_ commits, if not necessarily _one_ commit. That is to say: if you feel it's important for there to be more than one commit to help keep the history more clear and trackable, do a rebase and squash to a history you feel is solid and we can merge that _without_ simply squashing it all into one.

**Comment by dprotaso** [comment]
I have no preference - if you can squash and merge into a single commit
it’ll save me the trouble of rebasing

**Comment by dprotaso** [comment]
The test waits for all gateways to be `Accepted` prior to getting the Gateway and updating it.

The update can fail if the resource is modified after we fetch it _but_ before we perform the update - this is just regular K8s semantics.

My intention of the test is that the gateway resource is stable prior to doing the `Get`/`Update` - it may make more sense that we wait on another condition rather than `Accepted`.

Can you confirm what the exact mutation is?

**Pull request merged**
